main.tf

```diff
@@ -1 +1,98 @@
+# To have a NSG
+data "azurerm_network_security_group" "this" {
+  count               = length(var.nic_config.nsg_name) > 0 && length(var.nic_config.nsg_rg_name) > 0 ? 1 : 0
+  name                = var.nic_config.nsg_name
+  resource_group_name = var.nic_config.nsg_rg_name
+}
+ 
+resource "azurerm_public_ip" "this" {
+  count               = var.public_ip_config.enabled ? 1 : 0
+  name                = local.public_ip.name
+  resource_group_name = var.resource_group_name
+  location            = var.virtual_machine_config.location
+  allocation_method   = var.public_ip_config.allocation_method
 
+  lifecycle {
+    ignore_changes = [
+      tags
+    ]
+  }
+}
+
+resource "azurerm_network_interface" "this" {
+  name                = local.nic.name
+  location            = var.virtual_machine_config.location
+  resource_group_name = var.resource_group_name
+  dns_servers         = var.nic_config.dns_servers
+
+  ip_configuration {
+    name                          = local.nic.ip_config_name
+    subnet_id                     = var.subnet.id
+    private_ip_address_allocation = var.nic_config.private_ip == null ? "Dynamic" : "Static"
+    private_ip_address            = var.nic_config.private_ip
+    public_ip_address_id          = var.public_ip_config.enabled ? azurerm_public_ip.this[0].id : null
+  }
+
+  lifecycle {
+    ignore_changes = [
+      tags
+    ]
+  }
+}
+
+resource "azurerm_network_interface_security_group_association" "this" {
+  count                     = length(var.nic_config.nsg_name) > 0 ? 1 : 0
+  network_interface_id      = azurerm_network_interface.this.id
+  network_security_group_id = data.azurerm_network_security_group.this[0].id
+}
+
+resource "azurerm_linux_virtual_machine" "this" {
+  name                            = local.virtual_machine.name
+  computer_name                   = var.virtual_machine_config.hostname
+  location                        = var.virtual_machine_config.location
+  resource_group_name             = var.resource_group_name
+  size                            = var.virtual_machine_config.size
+  provision_vm_agent              = true
+  admin_username                  = var.virtual_machine_config.admin_username
+  admin_password                  = var.admin_password
```

Comment:
Right now it would be possible to specify an admin password and an ssh key. This would cause the module disable password auth. But when a password is set disable_password_authentication must be false (https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_virtual_machine#admin_password). So it would probably cause an error.

I would use variable validation to makes sure only one thing can be specified. This would give a better error message

Reply:
Fixed with a new variable map and somes validation